### PR TITLE
ConnectBlock: don't serialize block hash twice

### PR DESCRIPTION
**Pull request by jb55**
In the validation:block_connected tracepoint, we call block->GetHash(), which
ends up calling CBlockHeader::GetHash(), executing around 8000 serialization
instructions. We don't need to do this extra work, because block->GetHash() is
already called further up in the function. Let's save that value as a local
variable and re-use it in our tracepoint so there is no unnecessary tracepoint
overhead.

Shave off an extra 100 or so instructions from the validation:block_connected
tracepoint by reusing a nearby GetTimeMicros(). This brings the tracepoint down
to 54 instructions.  Still high, but much better than the previous ~154 and
8000 instructions which it was originally.

Signed-off-by: William Casarin <jb55@jb55.com>

### CODE REVIEW

**Comment by jb55** [comment]
cc @laanwj @0xB10C

**Comment by 0xB10C** [comment]
Concept ACK, will review in the next days.

ref: https://github.com/bitcoin/bitcoin/pull/23724

**Review by theStack — APPROVED** [review]
Code-review ACK b60c8fdc0772e78cb4c8b937f35afb3e25d3df56

Not related to the tracepoint overhead, but maybe still worth mentioning: note that in the PR branch, `block.GetHash()` is still called more than once in `CChainState::ConnectBlock`, as there is another invocation at the very start:
https://github.com/bitcoin/bitcoin/blob/a1e04e10799791a728cce4430808517598c8742c/src/validation.cpp#L1869
(As far as I know, we enforce compiling with assertions on, i.e. NDEBUG never being set)

**Comment by jb55** [comment]
On Sat, Jan 01, 2022 at 01:55:29PM -0800, Sebastian Falbesoner wrote:
>Not related to the tracepoint overhead, but maybe still worth mentioning: note that in the PR branch, `block.GetHash()` is still called more than once in `CChainState::ConnectBlock`, as there is another invocation at the very start:
>https://github.com/bitcoin/bitcoin/blob/a1e04e10799791a728cce4430808517598c8742c/src/validation.cpp#L1869
>(As far as I know, we enforce compiling with assertions on, i.e. NDEBUG never being set)

great catch. I've moved the blockHash variable to above the assert, so we shave off ~16,000 instructions now.

**Review by theStack — APPROVED** [review]
Code review re-ACK 8f8b1969e9d15117fe89ecc377d83e3f7b8dc5c1

Slightly off-topic, but could you elaborate how exactly you measure the number of saved instructions (and on which architecture)? Do you statically analyse the compiled binary or do it dynamically via `perf` or alike?

**Comment by jb55** [comment]
On Tue, Jan 04, 2022 at 10:57:28AM -0800, Sebastian Falbesoner wrote:
>Slightly off-topic, but could you elaborate how exactly you measure the number of saved instructions (and on which architecture)? Do you statically analyse the compiled binary or do it dynamically via `perf` or alike?

See https://github.com/bitcoin/bitcoin/pull/23724#issuecomment-996919963 for the method
and https://github.com/bitcoin/bitcoin/pull/23724#issuecomment-997455210 for where I determined that GetHash hits 8000 instructions.

**Review by luke-jr — APPROVED** [review]
utACK

I'm surprised we're not caching block hashes like we do for transactions, though?

**Comment by jb55** [comment]
On Tue, Jan 11, 2022 at 03:11:55PM -0800, Luke Dashjr wrote:
>I'm surprised we're not caching block hashes like we do for transactions, though?

Where is it cached for transactions? I see `CMutableTransaction::GetHash` but it looks roughly the same as `CBlockHeader::GetHash`.

**Comment by luke-jr** [comment]
Check `CTransaction`

**Comment by jb55** [comment]
On Tue, Jan 11, 2022 at 05:03:06PM -0800, Luke Dashjr wrote:
>Check `CTransaction`

gotcha, I see ComputeHash in the constructor. Yes it looks like CBlockHeader doesn't have caching, I'm guessing because it's a mutable thing?

**Comment by MarcoFalke** [comment]
Looks like this is also refactoring `ConnectBlock`, which is not tracing related? -> Renamed pull title

File: src/validation.cpp [hunk 1]
```diff
@@ -1866,7 +1866,10 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
 {
     AssertLockHeld(cs_main);
     assert(pindex);
-    assert(*pindex->phashBlock == block.GetHash());
+
+    uint256 blockHash = block.GetHash();
```
Comment (MarcoFalke) [inline src/validation.cpp:1870]:
```suggestion
    const uint256 block_hash{block.GetHash()};
```

style nit

**Review by maflcko — APPROVED** [review]
lgtm. Left a style nit

**Review by theStack — APPROVED** [review]
re-ACK eb8b22d5176d7abc6f93b4473df446105ca595e6

**Comment by 0xB10C** [comment]
ACK eb8b22d5176d7abc6f93b4473df446105ca595e6

I'm measuring 7 instructions. Code changes look good to me. [Still needs refactoring in the PR title](https://github.com/bitcoin/bitcoin/pull/23819#issuecomment-1010738359) :)

```log
(gdb) b src/validation:2161
Breakpoint 1 at 0x2c57b0: file src/validation.cpp, line 2161.
(gdb) b src/validation:
Breakpoint 2 at 0x2c57de: file src/validation.cpp, line 2172.
(gdb) run
[..]

Thread 22 "b-msghand" hit Breakpoint 1, CChainState::ConnectBlock (this=0x555555d19980, block=..., state=...,
    pindex=<optimized out>, view=..., fJustCheck=false) at validation.cpp:2161
2161	    LogPrint(BCLog::BENCH, "    - Index writing: %.2fms [%.2fs (%.2fms/blk)]\n", MILLI * (nTime5 - nTime4), nTimeIndex * MICRO, nTimeIndex * MILLI / nBlocksTotal);
(gdb) n
2163	    TRACE6(validation, block_connected,
(gdb) record btrace
(gdb) c
Continuing.

Thread 22 "b-msghand" hit Breakpoint 2, CCheckQueueControl<CScriptCheck>::~CCheckQueueControl (
    this=<optimized out>, __in_chrg=<optimized out>) at validation.cpp:2172
2172	    return true;
(gdb) record info
Undefined record command: "info".  Try "help record".
(gdb) info record
[..]
Recorded 7 instructions in 1 functions (0 gaps) for thread 22 (Thread 0x7fff2e7fc640 (LWP 508792)).
```

@jb55 you mention 54 instructions in eb8b22d5176d7abc6f93b4473df446105ca595e6. I think your measurements include the instructions from line 2161. I'm starting the measurement in line 2163.

https://github.com/bitcoin/bitcoin/blob/eb8b22d5176d7abc6f93b4473df446105ca595e6/src/validation.cpp#L2161-L2172

**Comment by laanwj** [comment]
Code review ACK eb8b22d5176d7abc6f93b4473df446105ca595e6

**Pull request merged**